src/ontodia/workspace/workspace.ts

```diff
@@ -194,51 +194,78 @@ export class Workspace extends Component<WorkspaceProps, State> {
         this.markup.paperArea.showIndicator(promise);
     }
 
-    forceLayout = () => {
+    private forceLayoutElements = (elements: Element[], group?: Element) => {
+        elements.forEach(element => {
+            const nestedNodes = this.model.elements.filter(el => el.group === element.id);
+
+            if (!nestedNodes.length) {
+                return;
+            }
+
+            this.forceLayoutElements(nestedNodes, element);
+        });
+
         const nodes: LayoutNode[] = [];
         const nodeById: { [id: string]: LayoutNode } = {};
-        for (const element of this.model.elements) {
+        for (const element of elements) {
             const {x, y, width, height} = boundsOf(element);
             const node: LayoutNode = {id: element.id, x, y, width, height};
             nodeById[element.id] = node;
             nodes.push(node);
         }
 
-        type LinkWithReference = LayoutLink & { link: Link };
-        const links: LinkWithReference[] = [];
+        const links: LayoutLink[] = [];
         for (const link of this.model.links) {
-            if (!this.model.isSourceAndTargetVisible(link)) { continue; }
+            if (!this.model.isSourceAndTargetVisible(link)) {
+                continue;
+            }
             const source = this.model.sourceOf(link);
             const target = this.model.targetOf(link);
-            links.push({
-                link,
-                source: nodeById[source.id],
-                target: nodeById[target.id],
-            });
+
+            const sourceNode = nodeById[source.id];
+            const targetNode = nodeById[target.id];
+
+            if (!sourceNode || !targetNode) {
+                continue;
```

Comment:
This flow control is kinda backwards here. To make it easier to follow it should be enough to invert the condition of `if` statement here and replace the `continue;` with `links.push(...)`.